api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,723 @@
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         map[string][]scaler.ResourceConsumer
 }
 
-func NewEnvironmentScaler() *EnvironmentScaler {
-	return &EnvironmentScaler{}
+func NewEnvironmentScaler(a *awsc.Client, e provider.EnvironmentProvider, s provider.ServiceProvider, t provider.TaskProvider, j job.Store, c config.APIConfig) *EnvironmentScaler {
+	return &EnvironmentScaler{
+		Client:              a,
+		EnvironmentProvider: e,
+		ServiceProvider:     s,
+		TaskProvider:        t,
+		JobStore:            j,
+		Config:              c,
+		// deployCache: map[string][]*Resource,
+	}
 }
 
 func (e *EnvironmentScaler) Scale(environmentID string) error {
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceConsumers, resourceProviders, err := e.GetCurrentState(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceProviders, unusedProviders, calcErrs, err := e.CalculateOptimizedState(clusterName, resourceConsumers, resourceProviders)
+	if err != nil {
+		return err
+	}
+
+	desiredScale := len(resourceProviders) - len(unusedProviders)
+	if err := e.ScaleToState(clusterName, desiredScale, unusedProviders); err != nil {
+		return err
+	}
+
+	if len(calcErrs) > 0 {
+		return errors.MultiError(calcErrs)
+	}
+
+	return nil
+}
+
+func (e *EnvironmentScaler) GetCurrentState(clusterName string) ([]scaler.ResourceConsumer, []*scaler.ResourceProvider, error) {
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	humanReadableProviders := ""
+	for _, r := range resourceProviders {
+		humanReadableProviders += fmt.Sprintf("\n    %#v", r)
+	}
+
+	log.Printf("[DEBUG] [EnvironmentScaler] resourceProviders for env '%s': %s", clusterName, humanReadableProviders)
+
+	var resourceConsumers []scaler.ResourceConsumer
+
+	serviceConsumers, err := e.getResourceConsumers_PendingServices(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resourceConsumers = append(resourceConsumers, serviceConsumers...)
+
+	ecsTaskConsumers, err := e.getResourceConsumers_TasksInECS(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resourceConsumers = append(resourceConsumers, ecsTaskConsumers...)
+
+	jobTaskConsumers, err := e.getResourceConsumers_TasksInJobs(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resourceConsumers = append(resourceConsumers, jobTaskConsumers...)
+
+	log.Printf("[DEBUG] [EnvironmentScaler] resourceConsumers for env '%s': %#v", clusterName, resourceConsumers)
+
+	return resourceConsumers, resourceProviders, nil
+}
+
+func (e *EnvironmentScaler) CalculateOptimizedState(clusterName string, resourceConsumers []scaler.ResourceConsumer, resourceProviders []*scaler.ResourceProvider) ([]*scaler.ResourceProvider, []*scaler.ResourceProvider, []error, error) {
```

Comment:
Apologies if I'm stepping on the toes of another suggestion, but this function seems like it's both taking and returning too much. Why return `[]errror` and `error`, for example? My one-minute assumption is that `[]errors` are multierrors being passed around?

Reply:
I'm not sure what it will look like yet, but this is something that will be addressed when I refactor things a bit to properly handle multi-errors.

Reply:
Overhauled error handling in this part of the code. ✅